File: web/packages/contracts/src/ws.ts
// WebSocket contract — TS mirror of server/app/schemas/ws.py (§5.3).

export type WSEventType =
  | 'unlock_success'
  | 'resource_credited'
  | 'entropy_update'
  | 'node_status_update'
  | 'bounty_new'
  | 'bounty_completed'
  | 'purity_prompt';

export interface UnlockSuccess {
  txn_id: string;
  node_id: string;
}

export interface ResourceGains {
  protein: number;
  carbs: number;
  lipids: number;
}

export interface ResourceCredited {
  txn_id: string;
  gains: ResourceGains;
  xp: number;
}

export interface EntropyUpdate {
  total_entropy: number;
  node_entropies: Record<string, number>;
}

export interface NodeStatusUpdate {
  node_id: string;
  status: string;
  health_avg: number;
}

export interface BountyNew {
  bounty: Record<string, unknown>;
}

export interface BountyCompleted {
  bounty_id: number;
}

export interface PurityPrompt {
  transaction_id: number;
}

// Tagged envelope every server->client message uses.
export interface WSEnvelope<T = unknown> {
  type: WSEventType;
  data: T;
}

// Discriminated map for type-narrowing on the client.
export interface WSEventMap {
  unlock_success: UnlockSuccess;
  resource_credited: ResourceCredited;
  entropy_update: EntropyUpdate;
  node_status_update: NodeStatusUpdate;
  bounty_new: BountyNew;
  bounty_completed: BountyCompleted;
  purity_prompt: PurityPrompt;
}

// Client -> server.
export type WSClientMessage =
  | { type: 'subscribe'; channels: string[] }
  | { type: 'ping' };

File: server/app/schemas/ws.py
"""WebSocket event contract — §5.3.

Every Server -> Client event is a tagged envelope: {"type": <event>, "data": {...}}.
Client -> Server messages are subscribe / ping.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel


# ---- Server -> Client payloads ----
class UnlockSuccess(BaseModel):
    txn_id: str
    node_id: str


class ResourceGains(BaseModel):
    protein: float = 0.0
    carbs: float = 0.0
    lipids: float = 0.0


class ResourceCredited(BaseModel):
    txn_id: str
    gains: ResourceGains
    xp: int


class EntropyUpdate(BaseModel):
    total_entropy: float
    node_entropies: dict[str, float]


class NodeStatusUpdate(BaseModel):
    node_id: str
    status: str
    health_avg: float


class BountyNew(BaseModel):
    bounty: dict[str, Any]


class BountyCompleted(BaseModel):
    bounty_id: int


class PurityPrompt(BaseModel):
    transaction_id: int


# Event type registry — keep in sync with frontend `web/packages/contracts`.
WSEventType = Literal[
    "unlock_success",
    "resource_credited",
    "entropy_update",
    "node_status_update",
    "bounty_new",
    "bounty_completed",
    "purity_prompt",
]


class WSEnvelope(BaseModel):
    type: WSEventType
    data: dict[str, Any]


# ---- Client -> Server messages ----
class WSSubscribe(BaseModel):
    type: Literal["subscribe"]
    channels: list[str]


class WSPing(BaseModel):
    type: Literal["ping"]
